torchgen/autoheuristic/train_decision.py

```diff
@@ -0,0 +1,604 @@
+# mypy: ignore-errors
+
+import itertools
+import json
+import math
+import warnings
+
+
+warnings.filterwarnings(
+    "ignore",
+    message="The behavior of DataFrame concatenation with empty or all-NA entries is deprecated",
+)
+
+import numpy as np
+import pandas as pd  # type: ignore[import-untyped]
+from scipy.stats import gmean
+from sklearn.model_selection import train_test_split
+from sklearn.tree import DecisionTreeClassifier
+from train import AHTrain
+
+
+class AHTrainDecisionTree(AHTrain):
+    def __init__(self):
+        super().__init__()
+
+    def get_time(self, row, choice):
+        choices_feedback = json.loads(row["choice2time"])
+        return choices_feedback.get(choice, None)
+
+    def top_k_classes(self, model, probas, k, avail_choices):
+        # Get classes and their corresponding probabilities
+        classes = model.classes_
+        class_proba_pairs = list(zip(classes, probas))
+
+        # Sort by probability (descending) and filter out zero probabilities
+        sorted_classes = [
+            c
+            for c, p in sorted(zip(classes, probas), key=lambda x: x[1], reverse=True)
+            if p > 0 and c in avail_choices
+        ]
+
+        # Return top k choices
+        return sorted_classes[:k]
+
+    def is_unsafe_leaf(self, row, predicted_config, choice2time):
+        return False
+
+    def get_unsafe_leaves(self, model, df, feature_columns):
+        X = df[feature_columns]
+        y = df["winner"]
+        leaf_ids = model.apply(X)
+        unique_leaves = np.unique(leaf_ids)
+
+        unsafe_leaves = []
+        for leaf in unique_leaves:
+            leaf_mask = leaf_ids == leaf
+            leaf_X = X[leaf_mask]
+
+            predicted_config = model.predict(leaf_X.iloc[[0]])[0]
+
+            for idx, row in leaf_X.iterrows():
+                choice2time = json.loads(df.loc[idx, "choice2time"])
+                if self.is_unsafe_leaf(row, predicted_config, choice2time):
+                    unsafe_leaves.append(leaf)
+                    break
+        return unsafe_leaves
+
+    def get_allowed_wrong_prediction_pct(self):
+        return 0.01
+
+    def get_grid_search_values(self):
+        return {
+            "max_depth": [5, 6, 7],
+            "min_samples_leaf": [1, 5, 10, 0.01, 0.05, 0.02],
+            "criterion": ["gini", "entropy"],
+        }
+
+    def predict(self, model, df, feature_columns):
+        predictions = model.predict(df[feature_columns])
+        proba = model.predict_proba(df[feature_columns])
+        leaf_ids = model.apply(df[feature_columns])
+        return predictions, proba, leaf_ids
+
+    def train_and_evaluate_models(
+        self, datasets, max_depths, min_samples_leafs, criterion_list, feature_columns
```

Comment:
Where do we pass in max_depths and how was it determined ?

Reply:
This method returns default values for the grid search https://github.com/pytorch/pytorch/pull/131613/files#diff-6a57055b20a319d0d47b571c2e286d3d3d95d7d0abcfc2d6109944947ddf28ebR90. I choose a maximum depth of 5, 6, and 7 because the learned heuristics do not perform well enough if the max depth is smaller. @Chillee and I have decided we want to have heuristics that are not too large, which is why 7 is the largest maximum depth that is tried out.

This method can also be overridden (see https://github.com/pytorch/pytorch/pull/131613/files#diff-64b6b146aecdb3f628198c363be9d64278f9b8bbb240134cff64488f391d434aR48)